Try/catch for 4.x

miripiruni opened the pull request:
See #298

Original task: https://nda.ya.ru/3RiZp2

miripiruni commented:
Benchmark results in web4 https://nda.ya.ru/3RiZii

miripiruni commented:
PR: https://nda.ya.ru/3RiZik

miripiruni commented on test/runtime-test.js:
```diff
@@ -947,7 +947,7 @@ describe('BEMHTML compiler/Runtime', function() {
     });
 
-    it('should throw error with one apply', function() {
+    xit('should throw error with one apply', function() {
```
The only question is: what shall I do with this cases? I think, as long as we introduce try/catch for block rendering it not valid anymore.

veged replied:
since it depend on `this.options.debug` there is should be two ways of tests: with throwing and without

miripiruni commented:
Review needed. cc @veged @zxqfox

St-Lukas commented on lib/bemhtml/runtime/index.js:
```diff
@@ -389,6 +390,32 @@ BEMHTML.prototype.runOne = function runOne(json) {
+        JSON.stringify(context.elemMods)
+      ].join(' ') +
+      '\n' +
+      e +
```
``e.stack``?

St-Lukas commented on lib/bemhtml/runtime/index.js:
```diff
@@ -389,6 +390,32 @@ BEMHTML.prototype.runOne = function runOne(json) {
+  } catch (e) {
+    console.error(e);
+
+    if (this.options.debug !== true)
```
Mode with throwing error will be useful for debugging.

miripiruni replied:
@St-Lukas `console.error(e)` it not useful for you?

St-Lukas replied:
Break on exceptions is very useful at debugging.
But with this changes you can't use it.

miripiruni commented:
Updated.

miripiruni commented on lib/bemhtml/runtime/index.js:
```diff
@@ -375,7 +375,8 @@ BEMHTML.prototype.runOne = function runOne(json) {
   }
 
-  var res = ent.run(context);
+  var res = this.tryRun(context, ent);
```
FOR Minor: if debug !== true ent.run
FOR Maj: this.tryRun

zxqfox commented on lib/bemhtml/runtime/index.js:
```diff
@@ -389,6 +390,36 @@ BEMHTML.prototype.runOne = function runOne(json) {
+  return res;
+};
+
+BEMHTML.prototype.buildRenderError = function buildRenderError(context, e) {
```
Why we decide to left this in production mode? Is it possible to drop this function in production?

miripiruni commented:
Updated. cc @tadatuta @veged 

I introduced `production` flag. If it’s value `true` bem-xjst will render all templates. If there is an error bem-xjst will log it via `console.error`.

```bash
$ cat stand.js
var bem = require('./');
var template = bem.compile(function() {
  block('page').attrs()(function() {
    var undef = applyNext();
    return undef.test = 'foo';
  });
}, { production: true });
var html = template.apply({ block: 'wrap', content: { block: 'page' } });
console.log(html);


$ node stand.js
BEMXJST ERROR: cannot render block page, elem undefined, mods {}, elemMods {} [TypeError: Cannot set property 'test' of undefined]
<div class="wrap"></div>
```

This way allows us to maintain backward capability. Also developers do not miss errors, since by default the program will stop its work. We can release it as minor.

miripiruni commented:
Notice that there is no documentation in v4.x. when I port this feature to v5 and higher I also add documentation.

miripiruni commented:
@veged @tadatuta review needed. Thank you!

veged commented on lib/bemhtml/runtime/index.js:
```diff
@@ -389,6 +392,25 @@ BEMHTML.prototype.runOne = function runOne(json) {
+  var res;
+
+  try {
+    res = ent.run(context);
```
you can `return` right here without unnecessary `res` variable

veged commented:
LGTM, except one comment

tadatuta commented on lib/bemhtml/runtime/index.js:
```diff
@@ -389,6 +392,25 @@ BEMHTML.prototype.runOne = function runOne(json) {
+  try {
+    res = ent.run(context);
+  } catch (e) {
+    console.error('BEMXJST ERROR: cannot render ' +
```
wouldn't it be more informative to show whole context:
```js
console.error('BEMXJST ERROR: cannot render ' + JSON.stringify(context, null, 4), e);
```

also it's possible to get rid of `res`:
```js
try {
    return ent.run(context);
} catch (e) {
    console.error(/* ... */);
}

return '';
```

miripiruni replied:
@tadatuta context contains circular structure and many information useless for debugging in this case.

St-Lukas commented on lib/bemhtml/runtime/index.js:
```diff
@@ -389,6 +392,21 @@ BEMHTML.prototype.runOne = function runOne(json) {
+        'elem ' + context.elem,
+        'mods ' + JSON.stringify(context.mods),
+        'elemMods ' + JSON.stringify(context.elemMods)
+      ].join(', '), e);
```
``e.stack || e``, please

miripiruni replied:
@St-Lukas e.stack is useless:


93:trycatch miripiruni$ node stand.js
BEMXJST ERROR: cannot render block page, elem undefined, mods {}, elemMods {} TypeError: Cannot set property 'test' of undefined
   at ContextChild.<anonymous> (evalmachine.<anonymous>:1863:19)
   at Match.tryCatch (evalmachine.<anonymous>:1240:15)
   at Match.exec (evalmachine.<anonymous>:1290:16)
   at Entity.defaultBody (evalmachine.<anonymous>:266:26)
   at Entity.run (evalmachine.<anonymous>:251:15)
   at BEMHTML.tryRun (evalmachine.<anonymous>:681:16)
   at BEMHTML.runOne (evalmachine.<anonymous>:663:10)
   at BEMHTML._run (evalmachine.<anonymous>:531:16)
   at BEMHTML.renderContent (evalmachine.<anonymous>:970:18)
   at BEMHTML.renderClose (evalmachine.<anonymous>:831:19)

miripiruni commented:
JFYI: Pulse test results: https://nda.ya.ru/3RiZii

miripiruni commented:
I think we can merge it. And port to 5.x, 6.x, 7.x.

veged commented:
yes

miripiruni commented:
Also in v5.2.0

miripiruni commented:
Also in v6.7.0

miripiruni commented:
Also in v7.4.0
